### PR TITLE
Make `LayerNorm` sample independent

### PR DESCRIPTION
**Pull request by jwa7**
Fix the implementation of `LayerNorm` and `InvariantLayerNorm` to calculate the mean and variance of blocks over all dimensions except the samples, such that transformations are applied to samples independently. The current implementation calculates these statistics over the samples, thus applying a BatchNorm-esque transformation.

# Contributor (creator of pull-request) checklist

 - [x] Tests updated (for new features and bugfixes)?
 - [x] Documentation updated (for new features)?
 - [ ] ~Issue referenced (for PRs that solve an issue)?~

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
 📚 [Download documentation preview for this pull-request](https://nightly.link/lab-cosmo/metatensor/actions/artifacts/1534277939.zip)

### CODE REVIEW

File: python/metatensor-learn/tests/layer_norm.py [hunk 1]
```diff
@@ -83,3 +83,19 @@ def test_equivariance(tensor, wigner_d_real):
     fRx = f(Rx)  # f(R . x)
 
     assert metatensor.allclose(fRx, Rfx, atol=1e-10, rtol=1e-10)
+
+
+def test_layer_norm_indep_samples():
```
Comment (Luthaf) [inline python/metatensor-learn/tests/layer_norm.py:88]:
```suggestion
def test_layer_norm_independent_samples():
```

**Review by Luthaf — COMMENTED** [review]
This looks alright, but it might be worth to get someone more familiar with LayerNorm to also check that this behaves as expected. Maybe @frostedoyster or @spozdn can help here?

**Comment by jwa7** [comment]
> The change is an improvement over the previous design, but I think the resulting models will not be equivariant

Yes in this case `LayerNorm` is not equivariance-preserving, but none of the base modules (i.e. `Tanh`, `Linear`) are in general. The `InvariantLayerNorm` is however, by applying a `LayerNorm` to invariants only, and the identity operator to covariants (in a similar fashion to `InvariantTanh`). 

> I think the correct thing to do would be to apply the layernorm just to the properties

@Luthaf what are your thoughts on this re: our discussion last week? As implemented we are contracting over components and properties at the moment when calculating the mean/std. I seem to remember there being a reason for not contracting over just properties - what was it again?

**Comment by frostedoyster** [comment]
Ah, nvm then, it looks good to me

**Comment by Luthaf** [comment]
> As implemented we are contracting over components and properties at the moment when calculating the mean/std. I seem to remember there being a reason for not contracting over just properties - what was it again?

So my idea was that when we start using this for covariant LayerNorm, we will set the mean to 0 (or ignore the mean in normalization); and we need to divide all components by the same value to preserve equivariance. So we need to compute the statistics over all properties and all components, producing only one standard deviation per sample (and not one per component)

**Comment by jwa7** [comment]
Makes sense - good as is then?

**Pull request merged**